### ISSUE TITLE
Proposal: Support cached properties

### ISSUE THREAD

**Issue by AlexSiepman** [dotnet/roslyn#524]
If you have a class  like this...

``` c#
    public class Sales
    {
        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
        {
            FoodPrices = foodPrices.ToList();
            NonFoodPrices = nonFoodPrices.ToList();

        }

        public IEnumerable<decimal> FoodPrices { get; }
        public IEnumerable<decimal> NonFoodPrices { get; }

        public decimal AveragePrice => FoodPrices.Concat(NonFoodPrices).Average();
    }
```

and you know that the value of AveragePrice never changes, you might want the property to be cached:

``` c#
    public class Sales
    {
        Lazy<decimal> _AveragePrice;

        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
        {
            FoodPrices = foodPrices.ToList();
            NonFoodPrices = nonFoodPrices.ToList();
            _AveragePrice = new Lazy<decimal>(() => FoodPrices.Concat(NonFoodPrices).Average());
        }

        public IEnumerable<decimal> FoodPrices { get; }
        public IEnumerable<decimal> NonFoodPrices { get; }

        public decimal AveragePrice => _AveragePrice.Value;
    }
```

Note: Because C#7 will provide better support for immutablity, I expect that more calculated properties will be based on non changing data. 
## Problem

If you cache those properties this way, it has the following disadvantages:
- You need to add the extra **boilerplate code** again and again. This makes the code harder to maintain / read.
- The **implementation** of the cached property now needs to be **outside** the method. This makes the code harder to read and maintain.
- You need even more boilerplate code if the method has **multiple lines of code**.
  
  AOP seems like an option but it creates non standard C# code and tools will not support it like standard C# code.
## Solution

Add new optional modifier to a property called **cached**:

``` c#
    public class Sales
    {
        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
        {
            FoodPrices = foodPrices.ToList();
            NonFoodPrices = nonFoodPrices.ToList();

        }

        public IEnumerable<decimal> FoodPrices { get; }
        public IEnumerable<decimal> NonFoodPrices { get; }

        public cached decimal AveragePrice => FoodPrices.Concat(NonFoodPrices).Average(); // New lazy modifier
    }
```

The compiler now will implement the boilerplate code. This solves the problem and also has the advantage that the caller of the property know the property will always return the same result.

When the property is no longer needed you simply remove the property and not removing the property, the constructor code and the field. When renaming the property, you do not have to rename the field.
## Implementation

Update after comment of @RichiCoder1: The implementation of:

``` c#
    public cached decimal AveragePrice => FoodPrices.Concat(NonFoodPrices).Average();
```

Needs to be something like this:

``` c#
    public class Sales
    {
        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
        {
            //...
            __lazy_AveragePrice = new Lazy(() => FoodPrices.Concat(NonFoodPrices).Average());
        }

        //...

        private Lazy<decimal> __lazy_AveragePrice;
        public decimal AveragePrice
        {
            get { return __lazy_AveragePrice.Value; }
        }
    }
```

**Note: This is just an example, as long as it is thread safe, an other implementation is also OK.**
## Methods

The same solution can also be applied to methods without parameters:

``` c#
    public cached decimal GetAveragePrice() // New lazy modifier
    {
        return FoodPrices.Concat(NonFoodPrices).Average();
    }
```

or even **with parameters** (see option of @bbarry ) 

Note: In the first version of this proposal the modifier was called **lazy**, but with **cached** it is now clear that it has a lazy implementation, not a lazy result.

**Comment by RichiCoder1** [comment]
So 
```csharp
public lazy decimal AveragePrice => FoodPrices.Concat(NonFoodPrices).Average();
```
Roughly boils down to something like this?
```csharp
public class Sales
{
        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
	{
		FoodPrices = foodPrices.ToList();
		NonFoodPrices = nonFoodPrices.ToList();
		__lazy_AveragePrice = new Lazy(() => FoodPrices.Concat(NonFoodPrices).Average());
	}
	
	public IEnumerable<decimal> FoodPrices { get; }
	public IEnumerable<decimal> NonFoodPrices { get; }
	
	private Lazy<decimal> __lazy_AveragePrice;
	public Lazy<decimal> AveragePrice
	{
		get { return __lazy_AveragePrice; }
	}
}
```
Or should that be opaque?
```csharp
public class Sales
{
        public Sales(IEnumerable<decimal> foodPrices, IEnumerable<decimal> nonFoodPrices)
	{
               //...
		__lazy_AveragePrice = new Lazy(() => FoodPrices.Concat(NonFoodPrices).Average());
	}
      
        //...
	
	private Lazy<decimal> __lazy_AveragePrice;
	public  decimal AveragePrice
	{
		get { return __lazy_AveragePrice.Value; }
	}
}
```

**Comment by AlexSiepman** [comment]
@RichiCoder1 Thanks, good comment! I would like it to be the opaque option, so the lazy modifier does *not* change the return type. I will update my proposal so that the implementation is clear.

**Comment by RichiCoder1** [comment]
Np! It does bring up the question, though, of whether or not the property should be annotated somehow to let reflection (at runtime) know it's a "lazy" property, or if that is simply opaque. Very edge case though, I'm not sure how much people would care.

**Comment by AlexSiepman** [comment]
@RichiCoder1 The only thing I really care about is that the property is only evaluated once. It is also nice that it is also clear to the caller and tools can now reason about it.

**Comment by RichiCoder1** [comment]
Yup. I was thinking about whether we should extend that ability to reason to the runtime.

This actually brings up another idea. You mentioned `lazy` on a method with no inputs. Why not expand that idea to introduce the idea of `memoized` methods, and removed the input limit?

**Comment by khellang** [comment]
How can you determine if the value has been created or not? If `Lazy<T>` would be an implementation detail and hidden from the developer, there's no `HasValue` property to check.

What about disposable resources? If there's no `HasValue` property, you'd have to actually call the (potentially expensive) `Value` property of the underlying `Lazy<T>` to get the instance, just to call its `Dispose` method?

**Comment by RichiCoder1** [comment]
@khellang I would imagine that the point would be to have it be opaque. So it'd be intentional not being able to access HasValue.

> What about disposable resources? If there's no HasValue property, you'd have to actually call the (potentially expensive) Value property of the underlying Lazy<T> to get the instance, just to call its Dispose method?

That's a good question, to which I'd have no good answer. Probably would frown on putting an Lazy <IDisposable> property in the first place.

**Comment by agocke** [comment]
What if you want it to be thread safe? And what if you want to use InterlockedExchange instead of locking?

**Comment by AlexSiepman** [comment]
@khellang If you need more control like the object that it returns implements IDisposable and you want only call the Dispose method if IsValueCreated is true, you need to implement the lazyness of the property yourself. 

I think in most of the C# projects 98% of the lazy properties has return values that does not Implement IDisposable or the Dispose method is not called. In general I think it is better to make it simple in the majority of the situations and hard for minority of the situations. The opposite would be: make it hard for all situations, because sometimes you need more control.

**Comment by AlexSiepman** [comment]
@agocke As uou can see in the implementation example, I want it to be thread safe (like the default of the Lazy<T> class). I searched for "new Lazy" and C# on the internet in general and more specific on Stackoverflow.com. After a few hundred search results, I found only one example of Lazy<T> that used the option to not be threadsafe. In that rare situation, you need to implement the lazyness of the property yourself.

**Comment by AlexSiepman** [comment]
@RichiCoder1 To be honest, when I suggested the lazy method with no parameters, I considered the option to remove the resrtiction "no parameters". The problem is that this is much harder to implement and it  raises many questions about equality etc. Also you do not need it as much as the simple option. This makes chance that this option is actually implemented so small, that I preferred the simple option that might have a chance. Maybe it is an option for C# 8 ;-)

**Comment by agocke** [comment]
@AlexSiepman And yet, all of the lazy code in the compiler uses InterlockedExchange, so there's clearly some use case for it :)

**Comment by AlexSiepman** [comment]
@agocke Good point! Maybe I should say that is is not used often in line of business apllications.

**Comment by Richiban** [comment]
I'd love this idea, but I think that use of Lazy<T> perhaps isn't necessary. The compiler could just insert the necessary so that the property's initialiser is not run in the containing class's constructor, but instead on first access of the property. Not hard to do!

**Comment by AlexSiepman** [comment]
@Richiban For me it is important that it is very easy to make a property lazy (thread safe). I do not care much about the implementation. This new option makes my application faster, it is easy to create and maintain and the intention is lear for the caller.

**Comment by bbarry** [comment]
@RichiCoder1, @AlexSiepman A method implementation for methods with parameters could do something like this:

    public lazy string FormalName(string first, string last, string title, string suffix) {
        return $"{title} {first} {last} {suffix}";
    }
    
becomes:

    public string FormalName(string first, string last, string title, string suffix)
    {
        return __value_FormalName.GetOrAdd(Tuple.Create(first, last, title, suffix), __impl_FormalName);
    }
    
    private ConcurrentDictionary<Tuple<string, string, string, string>, string> __value_FormalName = 
        new ConcurrentDictionary<Tuple<string, string, string, string>, string>();

    private string __impl_FormalName(Tuple<string, string, string, string> arguments)
    {
        var first = arguments.Item1;
        var last = arguments.Item2;
        var title = arguments.Item3;
        var suffix = arguments.Item4;
        return $"{title} {first} {last} {suffix}";
    }

I don't see any potential confusion on equality as far as parameters go (though if that was an `IEnumerable` or `IDisposable` return type there could be wierd stuff).

**Comment by chrisaut** [comment]
The proposal is that this would return a T, and not a Lazy\<T\>, correct?

Aren't we talking about Caching/Memorization at this point? What part about this is "Lazy"? 

To me, lazy implies that while I get "something" back that I can extract the value from, it isn't computed until it is really needed. That is, while a might call method/property x to get a Lazy\<T\>, store that in a local, I may later decided that hey, turns out I don't need it in which case the T would never be computed.

Wiki says:
```
In programming language theory, lazy evaluation, or call-by-need[1] is an evaluation strategy which
 delays the evaluation of an expression until its value is needed (non-strict evaluation) and which also
 avoids repeated evaluations (sharing).[2][3]
```

**Comment by AlexSiepman** [comment]
@chrisaut 
>The proposal is that this would return a T, and not a Lazy<T>, correct?

Yes.

>Aren't we talking about Caching/Memorization at this point? 
>What part about this is "Lazy"?

The implementation is lazy, not the return value. '**cached**' may be a better name for the new modifier the lazy. With "cached" it is clear the return value is not lazy. 

@chrisaut : Is 'cached' a better name for the new modifier? I might change/improve the proposal...

**Comment by chrisaut** [comment]
Yes I think cached would be better.

But I don't think this should be part of the signature at all. To a consumer, the fact that it's cached is kind of irrelevant, this is an implementation detail IMO.

Your proposed solution:
```C#
public lazy decimal AveragePrice => FoodPrices.Concat(NonFoodPrices).Average(); // New lazy modifier
```
But this is not much shorter than what you can do with C# already if you just create an ImplicitLazy implementation that just supports implicit operator to T and from Lazy or even Func. Yes, it's still a little ugly, however the Lazy constructor should support automatic type parameter ..... guessing or whatever the word is (so you don't have to say Lazy<decimal> anymore.). And you have getter only auto properties that you can assign your lazy too. Next I believe we will get implicit lamda conversion to Func<T>, in which case not much remains.

Basically today, all that remains is:
```C#
AveragePrice = new Lazy<decimal>(() => FoodPrices.Concat(NonFoodPrices).Average());
```
but I'm pretty sure by the time C#7 comes out, all that would remain is
```C#
AveragePrice = () => FoodPrices.Concat(NonFoodPrices).Average();
```

I did a little gist:
https://gist.github.com/chrisaut/e376cfae21511e15a44a

While I agree that caching like this is often nice, IMO this feature is not worth it at the language level, it can be implemented as a library. The language just needs to make it less painful to consume (and I think it does that).

**Comment by AlexSiepman** [comment]
@chrisaut and @RichiCoder1: I have update my proposal from 'lazy' to 'cached'. Thanks for your input.

@chrisaut: Your little gist is one of the best options so far, but still does has about the same disadvantages as I described in my proposal.
* The implementation is outside the property, If you go to the implemantation of the property, you don't see the implementation because it is implemented is somewhere else (in the constructor)
* The code is even harder to read/maintain if you have an implementation of the property that contains *multiple lines*.
* When the 'cached' modifier can also be applied to methods (with parameters), using a self created option needs even more boilerplate code.

You wrote: 

>  To a consumer, the fact that it's cached is kind of irrelevant, this is an implementation detail IMO.

It returns a normal value (not Lazy) and the return value is guaranteed to **return always the same value**. This could be relevant for the caller or for refactor tools.

**Comment by AlexSiepman** [comment]
@bbarry : I added your option to the proposal. This way I might ask to much and get nothing at all, but I like your option to much to ignore it...

**Comment by brandf** [comment]
I like the suggestion above about not using Lazy.  The point of laziness is to defer work until the property is accessed.  With the examples that use Lazy above, the ctor is allocating a Lazy every time.  If the compiler just did the laziness in a getter for the lazy method, then it would be effectively free unless accessed.

**Comment by Richiban** [comment]
I too support this idea, especially since C# is moving towards making immutable programming more easy. After all, if a collection can't change then you don't need to recompute the sum every time; the sum will never change if the collection can never change.

Essentially this:
```
public class MyClass
{
	public MyClass(IEnumerable<int> things)
	{
		Things = things;
	}
	
	public IEnumerable<int> Things { get; }
	public cached int Sum => Things.Sum();
}
```

would be the equivalent of writing this:
```
public class MySecondClass
{
	public MySecondClass(IEnumerable<int> things)
	{
		Things = things;
	}

	public IEnumerable<int> Things { get; }	
	
	private bool _sumEvaluated;
	private int _sum;
	public int Sum
	{
		get
		{
			if (!_sumEvaluated)
			{
				_sum = Things.Sum();
			}
				
			_sumEvaluated = true;
				
			return _sum;
		}
	}
}
```

Note that, since the compiler is free to generate as much boilerplate as it likes, it doesn't need to actually use the Lazy<T> class. Instead it would simply keep around a boolean flag to indicate whether or not the property has already been evaluated.

Note: I don't think this will work well at all with methods that accept parameters; the vast majority of developers would be surprised if `MyCachedMethod(1)` cached the same return value as `MyCachedMethod(2)`. If you don't do that, though, you have to start keeping around a Dictionary of cached values. And what about the tricky nature of equality in .NET? This is just too complicated. Best stick simply to properties for now.

**Comment by gafter** [comment]
We are now taking language feature discussion in other repositories:
- https://github.com/dotnet/csharplang for C# specific issues
- https://github.com/dotnet/vblang for VB-specific features
- https://github.com/dotnet/csharplang for features that affect both languages

Features that are under active design or development, or which are "championed" by someone on the language design team, have already been moved either as issues or as checked-in design documents. For example, the proposal in this repo "Proposal: Partial interface implementation a.k.a. Traits" (issue 16139 and a few other issues that request the same thing) are now tracked by the language team at issue 52 in https://github.com/dotnet/csharplang/issues, and there is a draft spec at https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md and further discussion at issue 288 in https://github.com/dotnet/csharplang/issues. Prototyping of the compiler portion of language features is still tracked here; see, for example, https://github.com/dotnet/roslyn/tree/features/DefaultInterfaceImplementation and issue 17952.

In order to facilitate that transition, we have started closing language design discussions from [the roslyn repo](https://github.com/dotnet/roslyn) with a note briefly explaining why. When we are aware of an existing discussion for the feature already in the new repo, we are adding a link to that. But we're not adding new issues to the new repos for existing discussions in this repo that the language design team does not currently envision taking on. Our intent is to eventually close the language design issues in the Roslyn repo and encourage discussion in one of the new repos instead.

Our intent is not to shut down discussion on language design - you can still continue discussion on the closed issues if you want - but rather we would like to encourage people to move discussion to where we are more likely to be paying attention (the new repo), or to abandon discussions that are no longer of interest to you.

If you happen to notice that one of the closed issues has a relevant issue in the new repo, and we have not added a link to the new issue, we would appreciate you providing a link from the old to the new discussion. That way people who are still interested in the discussion can start paying attention to the new issue.

Also, we'd welcome any ideas you might have on how we could better manage the transition. Comments and discussion about closing and/or moving issues should be directed to https://github.com/dotnet/roslyn/issues/18002. Comments and discussion about this issue can take place here or on an issue in the relevant repo.

-----------------

I am not moving this particular issue because I don't have confidence that the LDM would likely consider doing this.
